File: twitter-puddle/Puddle.java
/* 
 * CuGBabyBeaR  
 */

public class Puddle{

    public static void main(String[] args) {
        int [] testcase_1 = {2,5,1,2,3,4,7,7,6};
        int [] testcase_2 = {2,5,1,3,1,2,1,7,7,6};
        int [] testcase_3 = {6,1,4,6,7,5,1,6,4};

        System.out.println("Case testcase_1, total volume :" + calculate(testcase_1));
        System.out.println("Case testcase_2, total volume :" + calculate(testcase_2));
        System.out.println("Case testcase_3, total volume :" + calculate(testcase_3));
    }

    private static int calculate(int [] testcase){
        int p_l = 0;
        int p_r = testcase.length - 1;
        int max_l = testcase[p_l];
        int max_r = testcase[p_r];

        int volume = 0;
        while (p_r > p_l) {
            if (max_l < max_r){
                p_l++;
                if (testcase[p_l] >= max_l){
                    max_l = testcase[p_l];
                }else{
                    volume += max_l - testcase[p_l];
                }
            }else{
                p_r--;
                if (testcase[p_r] >= max_r){
                    max_r = testcase[p_r];
                }else{
                    volume += max_r - testcase[p_r];
                }
            }
        }

        return volume;
    }
}

File: twitter-puddle/puddle.cpp
/* 
 * CuGBabyBeaR  
 */

#include "stdio.h"

int volume_calcuate(int[],int);

int main(int argc, char const *argv[]){
    int testcase_1[9] = {2,5,1,2,3,4,7,7,6};
    int testcase_2[10] = {2,5,1,3,1,2,1,7,7,6};
    int testcase_3[9] = {6,1,4,6,7,5,1,6,4};

    printf("Case testcase_1 total volume : %i\n",volume_calcuate(testcase_1,9));
    printf("Case testcase_2 total volume : %i\n",volume_calcuate(testcase_2,10));
    printf("Case testcase_3 total volume : %i\n",volume_calcuate(testcase_3,9));
    return 0;
}

int volume_calcuate(int walls[] , int length){
    int * p_l = &walls[0];
    int * p_r = &walls[length-1];
    int max_l = walls[0];
    int max_r = walls[length-1];

    int total_volume = 0;

    while(p_r > p_l){
        if (max_l < max_r){
            p_l++;
            if (*p_l >= max_l){
                max_l = *p_l;
            }else{
                total_volume += max_l - *p_l;
            }
        }else{
            p_r--;
            if (*p_r >= max_r){
                max_r = *p_r;
            }
            else{
                total_volume += max_r - *p_r;
            }
        }
    }
    return total_volume;
}